### PR TITLE
1.10/1551 void voucher error message with success

### PR DESCRIPTION
**Pull request by katjam**
https://trello.com/c/6Ep7bSZL/1551-vouchers-that-have-already-been-paid-or-recorded-for-payment-cannot-be-marked-as-void-specify-voucher-in-error-message-eee-v

- Allow batches of void vouchers to contain non-voidable.
- success details contain list of unvoidable vouchers.

There is certainly room for improvement, this is a first hack.

### CODE REVIEW

File: app/Http/Controllers/Service/Admin/VouchersController.php [hunk 1]
```diff
@@ -53,26 +53,32 @@ public function void()
...
+        if ($voidableVouchers->count() === 0) {
+            // Whoops! None of the vouchers are voidable
             return redirect()
-                ->route('admin.vouchers.void')
-                ->withInput()
-                ->with('error_message', trans('service.messages.vouchers_batchtransition.blocked'));
+              ->route('admin.vouchers.void')
+              ->withInput()
+              ->with('error_message', trans('service.messages.vouchers_batchretiretransition.blocked'));
+            return;
```
Comment (AlinChindea) [inline app/Http/Controllers/Service/Admin/VouchersController.php:78]: if there are, shouldn't we get a success message and display it to the user?

File: app/Providers/CustomValidationProvider.php [hunk 2]
```diff
@@ -59,6 +59,34 @@ public function boot()
...
+            $val = Voucher::splitShortcodeNumeric($value);
+
+            // Grab the content of the parameter we pass the rule in a roundabout fashion
+            $otherVal = array_get(
+                $validator->getData(),
+                $parameters[0]
+            );
+
+            if (!empty($otherVal)) {
+                $other = Voucher::splitShortcodeNumeric($otherVal);
```
Comment (AlinChindea) [inline app/Providers/CustomValidationProvider.php:76]: we should be more descriptive with the variable names, no?

Comment (katjam) [reply app/Providers/CustomValidationProvider.php:76]: Was straight copy of the block above. If you decide to change, please change there also.

File: app/Voucher.php [hunk 3]
```diff
@@ -538,22 +442,51 @@ public static function getMinMaxVoucherDates(Collection $vouchers)
...
+    /**
+     * Gets a set of vouchers that are voidable.
+     *
+     * @param Builder $query
+     * @param array $states
+     * @return Builder
+     */
+    public function scopeInVoidableState(Builder $query)
+    {
+        $voidable_states = ['dispatched'];
```
Comment (AlinChindea) [inline app/Voucher.php:486]: shouldn't this include `printed`, too? Maybe am not fully getting it but above - line 180 you wrote that `$voidable_states = ['printed', 'dispatched'];`

File: app/Voucher.php [hunk 4]
```diff
@@ -174,28 +175,21 @@ public static function createRangeDefFromVoucherCodes($startCode, $endCode)
...
     }
 
-    /**
-     * Determines if the given voucher range contains entries already delivered
-     *
-     * @param $rangeDef object { 'start', 'end', 'shortcode', 'sponsor_id' }
-     * @return bool
-     */
-    public static function rangeIsDeliverable($rangeDef)
+    public static function collectionHasVoidable($currentstates)
```
Comment (AlinChindea) [inline app/Voucher.php:178]: this method doesn't seem to be used, neither are its argument `currentstates` and the declared variable (`$voidable_states`) 🤔

File: app/Voucher.php (hunk 3 above)
Comment (charlesstrange2) [reply app/Voucher.php:486]: We should only be voiding dispatched vouchers at this point

File: app/Http/Controllers/Service/Admin/DeliveriesController.php [hunk 5]
```diff
@@ -96,7 +96,8 @@ public function store(AdminNewDeliveryRequest $request)
                     DB::enableQueryLog();
```
Comment (charlesstrange2) [inline app/Http/Controllers/Service/Admin/DeliveriesController.php:96]: Whoops! this can go

File: app/Voucher.php (hunk 4 above)
Comment (charlesstrange2) [reply app/Voucher.php:178]: I'll prune it

File: app/Providers/CustomValidationProvider.php (hunk 2 above)
Comment (charlesstrange2) [reply app/Providers/CustomValidationProvider.php]: done

**Pull request merged**